### ISSUE TITLE
Weird behavior about schedule workflow

### ISSUE THREAD

**Issue by frozeNinK** [kubeflow/pipelines#3888]
The calculation of the next schedule time of a workflow is based on the prev workflow's schedule time, not max of schedule / start / completion time.

1. code below will chose the `lastJobEpoch` as the calculation base time:
https://github.com/kubeflow/pipelines/blob/58f1d132147dc2796e2ea69880316e98e4d258ae/backend/src/crd/controller/scheduledworkflow/util/cron_schedule.go#L51-L56

2. if configured with `no catch up`, code below will find the largest time which is before `now` and fits the cron job spec:
https://github.com/kubeflow/pipelines/blob/58f1d132147dc2796e2ea69880316e98e4d258ae/backend/src/crd/controller/scheduledworkflow/util/cron_schedule.go#L109-L119

3. `lastJobEpoch` is actually the job's schedule time instead of job end time:
https://github.com/kubeflow/pipelines/blob/58f1d132147dc2796e2ea69880316e98e4d258ae/backend/src/crd/controller/scheduledworkflow/controller.go#L512-L518
https://github.com/kubeflow/pipelines/blob/58f1d132147dc2796e2ea69880316e98e4d258ae/backend/src/crd/controller/scheduledworkflow/util/scheduled_workflow.go#L327

say the cron job is configured to run every day between 14:00 to 15:00, with max concurrency being 1
* a job (job A) is scheduled at 14:04, completed at 15:40
* job A's completion event arrives at KFP, trigger syncHandler
* KFP (say at time 15:44) will schedule a new job with schedule time 14:59 (due to above 1 & 2 & 3), well past the current time, 15:44 (with no catch up policy), which does not make sense.

**Comment by rmgogogo** [comment]
Thanks for reporting the issue. welcome PR if you already authored one.

**Comment by Bobgy** [comment]
/reopen
Sorry, I didn't quite follow the explanantion.
for example
> KFP (say at time 15:44) will schedule a new job with schedule time 14:59 (due to above 1 & 2 & 3), well past the current time, 15:44 (with no catch up policy), which does not make sense.

Can you rephrase in terms of, what happened and what is expected?
